lib/trino-filesystem-cache-alluxio/src/main/java/io/trino/filesystem/alluxio/ConsistentHashingCachingHostAddressProvider.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.filesystem.alluxio;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+import com.google.inject.Inject;
+import io.airlift.log.Logger;
+import io.trino.filesystem.cache.CachingHostAddressProvider;
+import io.trino.spi.HostAddress;
+import io.trino.spi.Node;
+import io.trino.spi.NodeManager;
+import jakarta.annotation.PostConstruct;
+import jakarta.annotation.PreDestroy;
+import org.ishugaliy.allgood.consistent.hash.ConsistentHash;
+import org.ishugaliy.allgood.consistent.hash.HashRing;
+import org.ishugaliy.allgood.consistent.hash.hasher.DefaultHasher;
+
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+
+public class ConsistentHashingCachingHostAddressProvider
+        implements CachingHostAddressProvider
+{
+    private static final Logger log = Logger.get(ConsistentHashingCachingHostAddressProvider.class);
+
+    private final NodeManager nodeManager;
+    private final ScheduledExecutorService hashRingUpdater = newSingleThreadScheduledExecutor(daemonThreadsNamed("hash-ring-refresher-%s"));
+
+    private final ConsistentHash<TrinoNode> consistentHashRing = HashRing.<TrinoNode>newBuilder()
+            .hasher(DefaultHasher.METRO_HASH)
+            .build();
+
+    @Inject
+    public ConsistentHashingCachingHostAddressProvider(NodeManager nodeManager)
+    {
+        this.nodeManager = requireNonNull(nodeManager, "nodeManager is null");
+    }
+
+    @Override
+    public List<HostAddress> getHosts(String splitPath)
+    {
+        return consistentHashRing.locate(splitPath)
```

Comment:
This would return just 1 or 0 nodes. Should we use `consistentHashRing.locate(splitPath, n)` instead where `n` can be configurable so that multiple HostAddress can be returned. In which case the splits on hot files/tables can be distributed well ?

Reply:
Suggested in https://github.com/trinodb/trino/pull/18719#discussion_r1352806444 as well. I don't think making it configurable is necessary at the moment.
The main issue with it is that when the "primary" node becomes less busy, we will go back to scheduling split there and leave the "secondary" node cache unused. It's still better than letting scheduler pick some other node randomly as we're doing currently.

Reply:
ping